### PR TITLE
refactor: refactor the UniversalFactory contract to LSP16

### PR DESCRIPTION
**Pull request by YamenMerhi**
## What does this PR introduce?
- Split the function `deployCreate2(..)` into `deployCreate2(..)` and `deployCreate2Init(..)` to allow more possibilities such as splitting the `msg.value` into the constructor and the initialize function
- Split the tests according to the three functions, and edge cases.
- Improve natspec

> The functions are split to take into consideration simplicity and runtime gas cost, any type of contract can be deployed with 1 function but we'll end up with many if else and ununsed parameters.

### CODE REVIEW

File: .github/workflows/build-lint-test.yml [hunk 1]
```diff
@@ -55,7 +55,7 @@ jobs:
             "lsp7",
             "lsp8",
             "lsp9",
-            "universalfactory",
+            "lsp16",
```
Comment (CJ42) [inline .github/workflows/build-lint-test.yml:58]: good catch 👌

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    ) public payable returns (address contractCreated) {
+        if (initializeCalldata.length == 0) revert ZeroInitializeCalldataNotAllowed();
+        if (constructorMsgValue + initializeCalldataMsgValue != msg.value)
+            revert InvalidMsgValueDistribution();
+
+        bytes32 generatedSalt = _generateSalt(initializeCalldata, providedSalt);
+        contractCreated = Create2.deploy(constructorMsgValue, generatedSalt, byteCode);
+        emit ContractCreated(contractCreated, providedSalt, true, initializeCalldata);
+
+        (bool success, bytes memory returndata) = contractCreated.call{
```
Comment (CJ42) [inline contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:134]: There is a linter error here.
We can discard it with solhint inline comments.

image

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        bool initializable = initializeCalldata.length > 0;
+        bytes32 generatedSalt = _generateSalt(initializeCalldata, providedSalt);
+
+        proxy = Clones.cloneDeterministic(baseContract, generatedSalt);
+        emit ContractCreated(proxy, providedSalt, initializable, initializeCalldata);
+
+        if (!initializable) {
+            if (msg.value > 0) revert ValueNotAllowedWithNonInitializableProxies();
+        } else {
+            (bool success, bytes memory returndata) = proxy.call{value: msg.value}(
```
Comment (CJ42) [inline contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:166]: Same here, linter error.

image

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+/**
+ * @dev reverts when msg.value sent to {deployCreate2Init} function is not equal to the sum of the
+ * `initializeCalldataMsgValue` and `constructorMsgValue`
+ */
+error InvalidMsgValueDistribution();
+
+/**
+ * @dev reverts when not passing any initializeCalldata to the {deployCreate2Init} function
+ */
+error ZeroInitializeCalldataNotAllowed();
```
Comment (CJ42) [inline contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:24]: _Suggested renaming options_

```suggestion
error InitializeCalldataRequired();
```

```suggestion
error MissingInitializeCalldata();
```

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+/**
+ * @dev reverts when sending value to the {deployCreate2Proxy} function if the contract being created
+ * is an uninitializable clone.
+ */
+error ValueNotAllowedWithNonInitializableProxies();
+
+/**
+ * @dev UniversalFactory contract can be used to deploy CREATE2 contracts; normal contracts and minimal
```
Comment (CJ42) [inline contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:33]: _Suggested changes_

Very good reference to Nick's article 👌 

```
/**
 * @dev UniversalFactory contract can be used to deploy normal or minimal proxy contracts (EIP-1167) using CREATE2.
 * This gives the ability to deploy the same contract at the same address on different chains.
 * If the contract has a constructor, the arguments will be part of the bytecode
 * If the contract has an `initialize` function, the parameters of this function will be included in
 * the salt to ensure that the parameters of the contract should be the same on each chain.
 *
 * This contract should be deployed using Nick's Method.
 * More information: https://weka.medium.com/how-to-send-ether-to-11-440-people-187e332566b7
 */
```

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol (hunk 4 above)
Comment (YamenMerhi) [reply contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:24]: Yeah, `InitializeCalldataRequired` is way more better, will do 👍

**Review by b00ste — APPROVED** [review]

File: contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        bytes calldata initializeCalldata
+    ) public payable returns (address proxy) {
+        bool initializable = initializeCalldata.length > 0;
+        bytes32 generatedSalt = _generateSalt(initializeCalldata, providedSalt);
+
+        proxy = Clones.cloneDeterministic(baseContract, generatedSalt);
+        emit ContractCreated(proxy, providedSalt, initializable, initializeCalldata);
+
+        if (!initializable) {
+            if (msg.value > 0) revert ValueNotAllowedWithNonInitializableProxies();
```
Comment (skimaharvey) [inline contracts/LSP16UniversalFactory/LSP16UniversalFactory.sol:169]: we could use `!=` (more gas efficient)

**Review by skimaharvey — APPROVED** [review]
looks great 🚀

**Review by CJ42 — APPROVED** [review]
